File: v2/core/src/prng.zig
/// Ratio: numerator/denominator probability. Allows very low probabilities
/// like 1-in-10-million that u8 percentages can't express.
pub const Ratio = struct {
    numerator: u32,
    denominator: u32,

    pub fn init(numerator: u32, denominator: u32) Ratio {
        return .{ .numerator = numerator, .denominator = denominator };
    }

    /// Zero probability (never triggers).
    pub fn zero() Ratio {
        return .{ .numerator = 0, .denominator = 1 };
    }

    /// Convenience: from percentage (e.g., 3 → 3/100).
    pub fn fromPercent(pct: u8) Ratio {
        return .{ .numerator = pct, .denominator = 100 };
    }

    pub fn isZero(self: Ratio) bool {
        return self.numerator == 0;
    }
};

/// Xorshift64 PRNG. Deterministic: same seed = identical sequence.
pub const Prng = struct {
    state: u64,

    pub fn init(seed: u64) Prng {
        return .{ .state = if (seed == 0) 1 else seed };
    }

    pub fn next(self: *Prng) u64 {
        var x = self.state;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.state = x;
        return x;
    }

    pub fn bounded(self: *Prng, bound: u64) u64 {
        return self.next() % bound;
    }

    pub fn intBounded(self: *Prng, comptime T: type, bound: T) T {
        return @intCast(self.bounded(@intCast(bound)));
    }

    /// Returns true with probability ratio.numerator / ratio.denominator.
    pub fn chance(self: *Prng, ratio: Ratio) bool {
        if (ratio.numerator == 0) return false;
        if (ratio.numerator >= ratio.denominator) return true;
        return self.bounded(ratio.denominator) < ratio.numerator;
    }
};

test "deterministic output" {
    var a = Prng.init(42);
    var b = Prng.init(42);
    var i: usize = 0;
    while (i < 1000) : (i += 1) {
        const std = @import("std");
        try std.testing.expectEqual(a.next(), b.next());
    }
}

test "ratio chance" {
    const std = @import("std");
    var p = Prng.init(999);

    // Zero ratio never fires
    var i: usize = 0;
    while (i < 100) : (i += 1) {
        try std.testing.expect(!p.chance(Ratio.zero()));
    }

    // 100% ratio always fires
    i = 0;
    while (i < 100) : (i += 1) {
        try std.testing.expect(p.chance(Ratio.init(1, 1)));
    }

    // 50% ratio fires roughly half the time
    var hits: u32 = 0;
    i = 0;
    while (i < 10000) : (i += 1) {
        if (p.chance(Ratio.init(1, 2))) hits += 1;
    }
    try std.testing.expect(hits > 4000 and hits < 6000);
}

test "bounded stays in range" {
    var p = Prng.init(123);
    var i: usize = 0;
    while (i < 1000) : (i += 1) {
        const std = @import("std");
        try std.testing.expect(p.bounded(10) < 10);
    }
}

File: worker/src/prng.rs
/// Ratio: numerator/denominator probability. Allows very low probabilities
/// like 1-in-10-million that u8 percentages can't express.
#[derive(Debug, Clone, Copy)]
pub struct Ratio {
    pub numerator: u32,
    pub denominator: u32,
}

impl Ratio {
    pub fn new(numerator: u32, denominator: u32) -> Self {
        Self {
            numerator,
            denominator,
        }
    }

    pub fn zero() -> Self {
        Self {
            numerator: 0,
            denominator: 1,
        }
    }

    pub fn from_percent(pct: u8) -> Self {
        Self {
            numerator: pct as u32,
            denominator: 100,
        }
    }

    pub fn is_zero(&self) -> bool {
        self.numerator == 0
    }
}

/// Xorshift64 PRNG. Identical algorithm to src/prng.zig in the
/// Zig control plane. Same seed produces identical sequence.
pub struct Prng {
    state: u64,
}

impl Prng {
    pub fn init(seed: u64) -> Self {
        Self {
            state: if seed == 0 { 1 } else { seed },
        }
    }

    pub fn next(&mut self) -> u64 {
        let mut x = self.state;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.state = x;
        x
    }

    pub fn bounded(&mut self, bound: u64) -> u64 {
        assert!(bound > 0);
        self.next() % bound
    }

    pub fn chance(&mut self, percent: u8) -> bool {
        self.bounded(100) < percent as u64
    }

    pub fn chance_ratio(&mut self, ratio: Ratio) -> bool {
        if ratio.numerator == 0 {
            return false;
        }
        if ratio.numerator >= ratio.denominator {
            return true;
        }
        self.bounded(ratio.denominator as u64) < ratio.numerator as u64
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deterministic_sequence() {
        let mut a = Prng::init(42);
        let mut b = Prng::init(42);
        for _ in 0..1000 {
            assert_eq!(a.next(), b.next());
        }
    }

    #[test]
    fn zero_seed_becomes_one() {
        let mut p = Prng::init(0);
        assert_ne!(p.next(), 0);
    }

    #[test]
    fn bounded_within_range() {
        let mut p = Prng::init(123);
        for _ in 0..1000 {
            assert!(p.bounded(10) < 10);
        }
    }

    #[test]
    fn chance_boundary() {
        let mut p = Prng::init(99);
        // chance(0) should always be false
        for _ in 0..100 {
            assert!(!p.chance(0));
        }
        // chance(100) should always be true
        let mut p2 = Prng::init(99);
        for _ in 0..100 {
            assert!(p2.chance(100));
        }
    }
}
